### PR TITLE
Remove redundant predicates after transitive closures

### PR DESCRIPTION
**Pull request by DevChattopadhyay**
Issue:
======
Current implementation of ORCA does not remove the redundant predicates after transitive closure.

Solution:
=========

This PR is trying to remove the redundant predicates based on the following steps.

- After the normalization step the predicates are already pushed down the tree. So they are redundant in the join condition.

- If the child of the join is a EopScalarCmp, we are not checking for redundancy because we need one child for the Hash join condition.

- If the child of the join is EopScalarBoolOp we are iterating through each child of EopScalarBoolOp and if its a EopScalarCmp with equlity type, we are checking if the value of that column is a constant.

- If it's a constant then it can be removed as it has been already pushed down the tree in the previous normalization step.
 
- If the redundant predicate consists of a distribution key then it's not removed.

- If a condition arises when all the childs are redundant then we are not removing all the childs as this will boil down to a nested loop. So in order to do a Hash join we are keeping one child even if its redundant based on if the column is a distribution key.

Setup:
======
create table foo(a text, b text);
create table bar(c text, d text);
explain select * from foo join bar on foo.a=bar.c and foo.b=bar.d where bar.d='cc';

Existing Behaviour:
===================
```
                            QUERY PLAN
-------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..862.00 rows=1 width=32)
   ->  Hash Join  (cost=0.00..862.00 rows=1 width=32)
         Hash Cond: ((foo.a = bar.c) AND (foo.b = bar.d))
         ->  Seq Scan on foo  (cost=0.00..431.00 rows=1 width=16)
               Filter: (b = 'cc'::text)
         ->  Hash  (cost=431.00..431.00 rows=1 width=16)
               ->  Seq Scan on bar  (cost=0.00..431.00 rows=1 width=16)
                     Filter: (d = 'cc'::text)
 Optimizer: Pivotal Optimizer (GPORCA)
(9 rows)
```

New Behaviour:
==============

```
                                  QUERY PLAN
-------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..862.00 rows=1 width=32)
   ->  Hash Join  (cost=0.00..862.00 rows=1 width=32)
         Hash Cond: (foo.a = bar.c)
         ->  Seq Scan on foo  (cost=0.00..431.00 rows=1 width=16)
               Filter: (b = 'cc'::text)
         ->  Hash  (cost=431.00..431.00 rows=1 width=16)
               ->  Seq Scan on bar  (cost=0.00..431.00 rows=1 width=16)
                     Filter: (d = 'cc'::text)
 Optimizer: Pivotal Optimizer (GPORCA)
(9 rows)
```

Co-authored-by: Hari krishna Maddileti [hmaddileti@vmware.com](mailto:hmaddileti@vmware.com)

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 1]
```diff
@@ -1597,6 +1597,156 @@ CExpressionPreprocessor::PexprScalarPredicates(
...
+					CRefCount::SafeRelease(pexprScalar);
+				}
+
+				// If the column attribute is not resolved to a constant value ,
+				// it means it's not redundant, so we can't remove it. And since
+				// it will be EopScalarCmp it can be used as a hash condition.
+				if (!canBeRemoved)
+				{
+					hashConditionPresent = true;
+					pexprChildOfBool->AddRef();
```
Comment (THANATOSLAVA) [inline src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:1661]: `pexprChildOfBool->AddRef();` can be moved after `CExpression *pexprChildOfBool = (*pexpr)[ul];`, cause it has to be appended to either `childrenArray` or `childrenRedundantArray`.

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 2]
```diff
@@ -1597,6 +1597,156 @@ CExpressionPreprocessor::PexprScalarPredicates(
...
+		// redundant array to perform a hash join.
+		if (0 == childrenArray->Size())
+		{
+			CExpression *pexprNewChild = (*childrenRedundantArray)[0];
+			pexprNewChild->AddRef();
+			CRefCount::SafeRelease(childrenArray);
+			CRefCount::SafeRelease(childrenRedundantArray);
+			return pexprNewChild;
+		}
+		else if (0 == childrenRedundantArray->Size())
```
Comment (THANATOSLAVA) [inline src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:1693]: Add `|| hashConditionPresent` to `else if`, and remove line 1714~1733, 1741
Also, can we replace line 1695~1709 with
```
pexpr->AddRef();
return pexpr;
```
which is case where none of the predicates is redundant, correct?

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 3]
```diff
@@ -1597,6 +1597,156 @@ CExpressionPreprocessor::PexprScalarPredicates(
...
+				}
+				else
+				{
+					COperator *pop = pexpr->Pop();
+					pop->AddRef();
+					CRefCount::SafeRelease(childrenRedundantArray);
+					return GPOS_NEW(mp) CExpression(mp, pop, childrenArray);
+				}
+			}
+			else
```
Comment (THANATOSLAVA) [inline src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:1732]: Can we add a comment for this case: when neither `childrenArray` or `childrenRedudantArray` is empty, we return `childrenArray` plus the one element of `childrenRedundantArray`.

**Review by THANATOSLAVA — COMMENTED** [review]
Good job. Could do with some refactoring and we will study the thoroughness.

**Comment by gpopt** [comment]
Closing this PR for now, as we have regression cases that we need to address first.

**Pull request closed without merging**